Improving performance, using caching when testing for primitives (#6252)

Boereck opened the pull request:
Caching primitive type names for faster lookup if a type is primitive.
Fixes #6252

jamesagnew commented:
@Boereck This is failing CI because of an android signature issue - We need to support android 26's specific weird hybrid of Java 11 and 17, which is annoying but usually not too hard.

In this case the failure appears to be `Collectors.toUnmodifiableSet()` which doesn't exist in A26 so you'll need to find another way of accomplishing that.

Boereck commented:
Thanks for the Info! I had a look at the Android API docs and updated the solution so to something that should (hopefully) work with API level 26. I also removed a call to `distinct()`, because I confused the API contracts of `Collectors.toMap()` and `Collectors.toSet()` before. Duplicate entries are not a problem for `Collectors.toSet()`.

Boereck commented:
@jamesagnew I added some test cases, since the covarage decreased a bit. Let me know if anything has to be improved.

jamesagnew reviewed (approved):
Looks good, thanks for the contribution!
